File: euler/primes.py
from collections.abc import Generator, Iterable
from math import sqrt


def prime_sieve(n: int) -> [bool]:
    """Get the sieve of primes below `n`

    Args:
        n: Index of last prime in list

    Returns:
        Boolean list with length n where index is number and boolean values is whether prime or not

    Examples:
        >>> prime_sieve(10)
        [False, False, True, True, False, True, False, True, False, False, False]
    """
    sieve = [True] * (n + 1)
    sieve[0] = sieve[1] = False
    for i in range(4, n + 1, 2):
        sieve[i] = False
    for i in range(3, int(sqrt(n)) + 1, 2):
        if sieve[i]:
            for j in range(i * i, n + 1, 2 * i):
                sieve[j] = False
    return sieve


def prime_numbers(n: int) -> Generator:
    """Generate prime numbers below `n`

    Args:
        n: Number below of which prime numbers will generate

    Yields:
        Next prime number in range from 0 to n - 1

    Examples:
        >>> print(list(prime_numbers(20)))
        [2, 3, 5, 7, 11, 13, 17, 19]
    """
    for i, prime in enumerate(prime_sieve(n)):
        if prime:
            yield i


def is_prime(n: int) -> bool:
    """Check if number is prime

    Args:
        n: Number

    Returns:
        True if number is prime, False otherwise

    Examples:
        >>> print(is_prime(23))
        True
        >>> print(is_prime(22))
        False
    """
    if n < 2:
        return False
    elif n < 4:
        return True
    elif n % 2 == 0 or n % 3 == 0:
        return False
    for i in range(6, int(sqrt(n)) + 2, 6):
        if n % (i - 1) == 0 or n % (i + 1) == 0:
            return False
    return True


def smallest_prime_factor(n: int, primes: Iterable = None) -> int:
    """Get smallest prime factor of number

    Args:
        n: Number
        primes: Generator of prime numbers below at least sqrt(n)

    Returns:
        Smallest prime number which proper divides n

    Examples:
        >>> print(smallest_prime_factor(23))
        23
        >>> n = 22
        >>> primes = list(prime_numbers(int(sqrt(n))))
        >>> print(smallest_prime_factor(n, primes))
        2
    """
    if primes is None:
        primes = prime_numbers(int(sqrt(n)))
    for prime in primes:
        if n % prime == 0:
            return prime
    return n

File: euler/R/primes.R
#' Get the sieve of primes below n
#' 
#' @param n Index of last prime in vector
#' @returns Boolean vector with length n where index is number and boolean values is whether prime or not
#' @examples
#' prime_sieve(10)
prime_sieve <- function(n) {
    sieve <- rep(TRUE, n)
    sieve[1] <- FALSE
    if (n < 4) return(sieve)
    sieve[seq(4, n, 2)] <- FALSE
    if (n >= 9) for (i in seq(3, as.integer(sqrt(n)), 2)) if (sieve[i]) sieve[seq(i * i, n, 2 * i)] <- FALSE
    return(sieve)
}

#' Get primes below n
#' 
#' @param n Number below of which prime numbers will generate
#' @returns Vector with prime numbers
#' @examples
#' prime_numbers(20)
prime_numbers <- function(n) which(prime_sieve(n))

#' Check if number is prime
#'
#' @param n Number
#' @returns TRUE if number is prime, FALSE otherwise
#' @examples
#' is.prime(23)
#' is.prime(22)
is.prime <- function(n) {
    if (n < 2) {
        return(FALSE)
    } else if (n < 4) {
        return(TRUE)
    } else if (n %% 2 == 0 | n %% 3 == 0) {
        return(FALSE)
    } else if (n < 25) {
        return(TRUE)
    }
    for (i in seq(6, as.integer(sqrt(n)) + 1, 6)) if (n %% (i - 1) == 0 || n %% (i + 1) == 0) return(FALSE)
    return(TRUE)
}

#' Get smallest prime factor of number
#' 
#' @param n Number
#' @param primes Vector of primes below at least sqrt(n)
#' @returns Smallest prime number which proper divides n
#' @examples
#' smallest_prime_factor(23)
#' n <- 22
#' primes <- prime_numbers(as.integer(sqrt(n)))
#' smallest_prime_factor(n, primes)
smallest_prime_factor <- function(n, primes = prime_numbers(as.integer(sqrt(n)))) {
    for (prime in primes) if (n %% prime == 0) return(prime)
    return(n)
}
